### PR TITLE
CB-12994 [AWS] Autoscale stack update failed after upgrading Data Hub

### PR DESCRIPTION
**Pull request by schfeca75**
* Instead of update the cloudformation template, we create a new version of LaunchTemplate for all the autoscale groups with the new launch template fields
* not only the imageid can be changeable
* new role permission is needed on aws: ec2:ModifyLaunchTemplate

See detailed description in the commit message.

### CODE REVIEW

File: cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    public void updateFields(AuthenticatedContext authenticatedContext, CloudResource cfResource,
+            Map<LaunchTemplateField, String> updatableFields, boolean dryRun) {
+        AwsCredentialView credentialView = new AwsCredentialView(authenticatedContext.getCloudCredential());
+        String regionName = authenticatedContext.getCloudContext().getLocation().getRegion().getRegionName();
+        AmazonCloudFormationClient cloudFormationClient = awsClient.createCloudFormationClient(credentialView, regionName);
+        AmazonAutoScalingClient autoScalingClient = awsClient.createAutoScalingClient(credentialView, regionName);
+        AmazonEc2Client ec2Client = awsClient.createEc2Client(credentialView, regionName);
+        Map<AutoScalingGroup, String> autoScalingGroups = autoScalingGroupHandler.getAutoScalingGroups(cloudFormationClient, autoScalingClient, cfResource);
+        autoScalingGroups = filterGroupsForDryRun(autoScalingGroups, dryRun);
+        for (Map.Entry<AutoScalingGroup, String> asgEntry : autoScalingGroups.entrySet()) {
```
Comment (keyki) [inline cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java:54]: Please add more log messages across the code.

**Review by keyki — APPROVED** [review]

File: cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        LOGGER.debug("Modifying the {} fields for the [{}] autoscaling groups' launchtemplates [dryrun: {}]", updatableFields, autoScalingGroups.values(),
+                dryRun);
+        for (Map.Entry<AutoScalingGroup, String> asgEntry : autoScalingGroups.entrySet()) {
+            AutoScalingGroup autoScalingGroup = asgEntry.getKey();
+            LOGGER.debug("Creating new launchtemplate version for [{}] autoscale group...", autoScalingGroup.getAutoScalingGroupName());
+            LaunchTemplateSpecification launchTemplateSpecification = getLaunchTemplateSpecification(autoScalingGroup);
+            CreateLaunchTemplateVersionResult createLaunchTemplateVersionResult = getCreateLaunchTemplateVersionRequest(ec2Client, updatableFields,
+                    launchTemplateSpecification);
+            modifyLaunchTemplate(ec2Client, launchTemplateSpecification, createLaunchTemplateVersionResult, !dryRun);
+            if (!dryRun) {
```
Comment (tiborpopovics) [inline cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java:62]: Can you please flip the `If` statement to eliminate the unnecessary negation?

File: cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    private Map<AutoScalingGroup, String> filterGroupsForDryRun(Map<AutoScalingGroup, String> autoScalingGroups, boolean dryRun) {
+        if (dryRun) {
+            Optional<Map.Entry<AutoScalingGroup, String>> firstGroup = autoScalingGroups.entrySet().stream().findFirst();
+            Map<AutoScalingGroup, String> result;
+            if (firstGroup.isPresent()) {
+                Map.Entry<AutoScalingGroup, String> firstGroupEntry = firstGroup.get();
+                result = Map.of(firstGroupEntry.getKey(), firstGroupEntry.getValue());
+            } else {
+                result = Map.of();
```
Comment (tiborpopovics) [inline cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsLaunchTemplateUpdateService.java:78]: Please use Collections.emptyMap()

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/event/ClusterUpgradeValidationHandlerSelectors.java [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 public enum ClusterUpgradeValidationHandlerSelectors implements FlowEvent {
 
+    VALIDATE_UPDATE,
```
Comment (tiborpopovics) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/event/ClusterUpgradeValidationHandlerSelectors.java:7]: Please rename this event to more expressive.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/CloudProviderUpdateCheckHandler.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        AuthenticatedContext ac;
+        StackEvent response;
+        try {
+            ac = connector.authentication().authenticate(request.getCloudContext(), request.getCloudCredential());
+            connector.resources().checkUpdate(ac, request.getCloudStack(), request.getCloudResources());
+            response = new ClusterUpgradeUpdateCheckSuccess(request.getResourceId(), request.getCloudStack(), request.getCloudCredential(),
+                    request.getCloudContext());
+        } catch (Exception e) {
+            String errorMessage = e.getMessage();
+            LOGGER.info(errorMessage, e);
```
Comment (tiborpopovics) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/CloudProviderUpdateCheckHandler.java:46]: Please change the log level to ERROR

**Review by tiborpopovics — APPROVED** [review]

**Pull request merged**